admin/test/test_release.py

```diff
@@ -661,3 +669,704 @@ def test_publish_non_release_fails(self):
+        expected_keys.update({
+            os.path.join(self.target_key, package): self.repo_contents[package]
+            for package in self.repo_contents})
+
+        self.assertDictContainsSubset(
+            expected_keys,
+            aws.s3_buckets[self.target_bucket])
+
+    def test_repository_metadata_index_updated(self):
+        """
+        Calling :func:`update_repo` updates the repository metadata index.
+        """
+        index_path = os.path.join(self.target_key, 'repodata', 'repomd.xml')
+        existing_s3_keys = {
+            index_path: 'old_metadata_index',
+        }
+
+        aws = FakeAWS(
+            routing_rules={},
+            s3_buckets={
+                self.target_bucket: existing_s3_keys.copy(),
+            },
+        )
+
+        self.update_repo(
+            aws=aws,
+            yum=FakeYum(),
+            rpm_directory=self.rpm_directory,
+            target_bucket=self.target_bucket,
+            target_key=self.target_key,
+            source_repo=self.create_fake_repository(files=self.repo_contents),
+            packages=self.packages,
+            flocker_version=self.dev_version,
+            distro_name=self.operating_systems[0]['distro'],
+            distro_version=self.operating_systems[0]['version'],
+        )
+
+        self.assertNotEqual(
+            existing_s3_keys[index_path],
+            aws.s3_buckets[self.target_bucket][index_path])
+
+    def test_existing_metadata_files_not_uploaded(self):
+        """
+        Calling :func:`update_repo` does not update repository metadata files
+        which are not the index.
+        """
+        index_path = os.path.join(self.target_key, 'repodata', 'repomd.xml')
+        existing_metadata_file = os.path.join(self.target_key, 'repodata',
+                                              'filelists.xml.gz')
+
+        existing_s3_keys = {
+            index_path: 'old_metadata_index',
+            existing_metadata_file: 'old_metadata_content',
+        }
+
+        aws = FakeAWS(
+            routing_rules={},
+            s3_buckets={
+                self.target_bucket: existing_s3_keys.copy(),
+            },
+        )
+
+        self.update_repo(
+            aws=aws,
+            yum=FakeYum(),
+            rpm_directory=self.rpm_directory,
+            target_bucket=self.target_bucket,
+            target_key=self.target_key,
+            source_repo=self.create_fake_repository(files=self.repo_contents),
+            packages=self.packages,
+            flocker_version=self.dev_version,
+            distro_name=self.operating_systems[0]['distro'],
+            distro_version=self.operating_systems[0]['version'],
+        )
+
+        # This tests that filelists.xml.gz, a metadata file which is not the
+        # index, still has the content 'old_metadata_content', even though
+        # CreateRepo changes this content. This shows that this file which
+        # already existed on S3 is not re-uploaded.
```

Comment:
I think this isn't something we care about (uploading a metadata file twice isn't a big issue), but we *don't* want to overwrite existing differently named (due to hashses) metadata.